### PR TITLE
Calculate GPU/CPU frame times in OrbitFakeClient

### PR DESCRIPTION
**Pull request by npcdoom**
Add new functionality to the FakeClient that will allow
measuring the GPU/CPU frame times using the GPU submission
events and instrumented QueuePresent calls from ggpvlk.so

Bug: http://b/204173805
Test: Automated/manuals runs of benchmark applications.

### CODE REVIEW

**Comment by dpallotti** [comment]
Adding @florian-kuebler as he wrote the Vulkan layer.

File: src/FakeClient/FakeClientMain.cpp
Comment (dpallotti) [inline src/FakeClient/FakeClientMain.cpp:398]:
```suggestion
    // Some application don't call libVulkan library directly; instead, they just query the
```

File: src/FakeClient/FakeClientMain.cpp
Comment (dpallotti) [inline src/FakeClient/FakeClientMain.cpp:399]:
```suggestion
    // function addresses and use those. So let's just instrument the layer
```

File: src/FakeClient/FakeClientMain.cpp
Comment (dpallotti) [inline src/FakeClient/FakeClientMain.cpp:437]: This case should be a failure.

**Review by dpallotti — COMMENTED** [review]
Part 1. Need to look at `GraphicsCaptureEventProcessor.h`.

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (dpallotti) [inline src/FakeClient/GraphicsCaptureEventProcessor.h:35]:
```suggestion
// This implementation of CaptureEventProcessor mostly discards all events it receives, except for
```

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (dpallotti) [inline src/FakeClient/GraphicsCaptureEventProcessor.h:52]: ProcessGPUQueueSubmission: GPU -> Gpu
CalculateGPUTimes: GPU -> Gpu
CalculateCPUTimes: CPU -> Cpu
WriteToCSVFile: CSV -> Csv
kCPUFrameTimeFilename: CPU -> Cpu
kGPUFrameTimeFilename: GPU -> Gpu
kCSVHeader: CSV -> Csv

Styleguide: "prefer to capitalize abbreviations as single words, e.g., StartRpc() rather than StartRPC()"
Yeah, exactly the opposite of golang...

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (dpallotti) [inline src/FakeClient/GraphicsCaptureEventProcessor.h:12]: I think that we tend to still nest anonymous namespaces inside the module's namespace (`orbit_fake_client`).

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (dpallotti) [inline src/FakeClient/GraphicsCaptureEventProcessor.h:126]: Why 1024? Since this is time, I expect something round like one second rather than a power of two.

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (dpallotti) [inline src/FakeClient/GraphicsCaptureEventProcessor.h:216]: Why 1024 buckets? Since this is time, I expect something round like one second rather than a power of two.

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (dpallotti) [inline src/FakeClient/GraphicsCaptureEventProcessor.h:66]: Either remove "that is possible", or "can".

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (dpallotti) [inline src/FakeClient/GraphicsCaptureEventProcessor.h:65]:
```suggestion
  // calls. The algorithm take into account that is possible that two jobs from different queues
```

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (dpallotti) [reply src/FakeClient/GraphicsCaptureEventProcessor.h:66]:
```suggestion
  // calls. The algorithm take into account that is possible that two jobs from different queue's
  // can overlap, so the time is not accounted for multiple times.
```

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (dpallotti) [reply src/FakeClient/GraphicsCaptureEventProcessor.h:82]: @florian-kuebler ? Otherwise it might be worthwhile to sort no matter what.

Comment (ronaldfw) [reply src/FakeClient/GraphicsCaptureEventProcessor.h:82]: For me the question arose partly because you are not sorting here, but you are sorting below, so as a reader I have to assume it's not necessary here but necessary below, but it's not obvious why. 

Considering from the perspective of someone looking at this code at a later time (maybe because of a bug, etc), I would make assumptions about sort order explicit and comment why are you (not) sorting, in particular, because these are external assumptions about a different component (the layer supplying the data). 

As examples, if you decide to not sort:
```
// We assume timestamps come in order, so we do not sort them. 
[...]
```

If you decide to sort:
```
// Timestamps should come in order, but we sort regardless to be careful and since runtime 
// doesn't matter that much here.
sort(...)
```

These comments make it obvious why you wrote the code the way you did and make it easy to change it if assumptions change (or we find out they are incorrect).

Comment (florian-kuebler) [reply src/FakeClient/GraphicsCaptureEventProcessor.h:82]: So the events are actually coming in order by CPU timestamp. But there is a [bit work done on the layer](https://github.com/google/orbit/blob/main/src/OrbitVulkanLayer/SubmissionTracker.h#L486), so it not obvious here.

I suggest to either add a comment or even just perform the sort here once as suggested.

**Review by florian-kuebler — APPROVED** [review]
I just briefly went through the code and only looked particularly at the computation part where I was mentioned.

Please wait until the others also approve (consider this a +1).

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (npcdoom) [reply src/FakeClient/GraphicsCaptureEventProcessor.h:82]: The reason I'm sorting the command buffer based on GPU times is because in Vulkan they can be executed out of order and also command buffers from different queues are able to run in parallel.

For the CPU timestamps the assumption is that they always be received in order

Comment (florian-kuebler) [reply src/FakeClient/GraphicsCaptureEventProcessor.h:82]: Thanks for the explanation. I suggest to add a comment here, that you assume the sorting (or just sort for consistency), and also add a comment below, that you need to sort based on what you said.

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (npcdoom) [reply src/FakeClient/GraphicsCaptureEventProcessor.h]: No particular reason, I maintained what the old tool had

File: src/FakeClient/FakeClientMain.cpp
Comment (florian-kuebler) [inline src/FakeClient/FakeClientMain.cpp:430]: No break is needed now.

File: src/FakeClient/FakeClientMain.cpp
Comment (npcdoom) [reply src/FakeClient/FakeClientMain.cpp]: Done

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (npcdoom) [reply src/FakeClient/GraphicsCaptureEventProcessor.h]: Done

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (npcdoom) [reply src/FakeClient/GraphicsCaptureEventProcessor.h]: Done

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (npcdoom) [reply src/FakeClient/GraphicsCaptureEventProcessor.h:250]: No particular reason, I maintained what the old tool had

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (npcdoom) [reply src/FakeClient/GraphicsCaptureEventProcessor.h]: Done

File: src/FakeClient/FakeClientMain.cpp
Comment (npcdoom) [reply src/FakeClient/FakeClientMain.cpp]: Done

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (npcdoom) [reply src/FakeClient/GraphicsCaptureEventProcessor.h:16]: Done

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (npcdoom) [reply src/FakeClient/GraphicsCaptureEventProcessor.h]: Done

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (npcdoom) [reply src/FakeClient/GraphicsCaptureEventProcessor.h]: Done

File: src/FakeClient/Flags.h
Comment (npcdoom) [reply src/FakeClient/Flags.h]: Done

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (npcdoom) [reply src/FakeClient/GraphicsCaptureEventProcessor.h:114]: Added the comment so its clear why it needs sorting

File: src/FakeClient/FakeClientMain.cpp
Comment (npcdoom) [reply src/FakeClient/FakeClientMain.cpp]: Done

File: src/FakeClient/FakeClientMain.cpp
Comment (npcdoom) [reply src/FakeClient/FakeClientMain.cpp]: Done

File: src/FakeClient/FakeClientMain.cpp
Comment (npcdoom) [reply src/FakeClient/FakeClientMain.cpp]: Removed the one below and kept this one

File: src/FakeClient/FakeClientMain.cpp
Comment (npcdoom) [reply src/FakeClient/FakeClientMain.cpp]: Done

File: src/FakeClient/FakeClientMain.cpp
Comment (npcdoom) [reply src/FakeClient/FakeClientMain.cpp]: Done

File: src/FakeClient/FakeCaptureEventProcessor.h
Comment (npcdoom) [reply src/FakeClient/FakeCaptureEventProcessor.h:45]: Done

File: src/FakeClient/FakeClientMain.cpp
Comment (npcdoom) [reply src/FakeClient/FakeClientMain.cpp]: Done

File: src/FakeClient/FakeCaptureEventProcessor.h
Comment (dpallotti) [reply src/FakeClient/FakeCaptureEventProcessor.h:45]: I guess this is still open.

File: src/FakeClient/FakeCaptureEventProcessor.h
Comment (dpallotti) [inline src/FakeClient/FakeCaptureEventProcessor.h:56]: Maybe just declare a single `std::filesystem::path file_path{absl::GetFlag(FLAGS_output_path)};` at the top of the destructor, and use `operator/` instead of `.append` where it's used.

File: src/FakeClient/FakeClientMain.cpp
Comment (dpallotti) [reply src/FakeClient/FakeClientMain.cpp:110]: Discussed offline, we are now instrumenting a different function.

File: src/FakeClient/FakeClientMain.cpp
Comment (dpallotti) [inline src/FakeClient/FakeClientMain.cpp:273]:
```suggestion
  LOG("Stopped watching \"%s\"", file_path);
```

File: src/FakeClient/FakeClientMain.cpp
Comment (dpallotti) [inline src/FakeClient/FakeClientMain.cpp:299]: I would declare a variable to hold the result of `absl::GetFlag(FLAGS_pid_file_path)`.

File: src/FakeClient/FakeClientMain.cpp
Comment (dpallotti) [inline src/FakeClient/FakeClientMain.cpp:391]: This comment needs to be updated. E.g.:
"We want to instrument vkQueuePresentKHR, but some applications don't call libVulkan library directly..." etc.

File: src/FakeClient/Flags.h
Comment (dpallotti) [inline src/FakeClient/Flags.h:19]: Here clang-tidy suggests you to avoid else after return.

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (dpallotti) [inline src/FakeClient/GraphicsCaptureEventProcessor.h:152]:
```suggestion
    // There's no guarantee that the submission order is mantained throughout the execution; command
```

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (dpallotti) [inline src/FakeClient/GraphicsCaptureEventProcessor.h:148]:
```suggestion
    // The frame gpu time is calculated as the union of all the command buffer intervals; to do this,
```

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (dpallotti) [inline src/FakeClient/GraphicsCaptureEventProcessor.h:163]:
```suggestion
      // If the interval doesn't overlap just add the length of the interval to the frame time, else
```

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (dpallotti) [inline src/FakeClient/GraphicsCaptureEventProcessor.h:164]:
```suggestion
      // if there's overlap just add the new part of the interval that hasn't yet been taken into
```

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (dpallotti) [inline src/FakeClient/GraphicsCaptureEventProcessor.h:256]:
```suggestion
  // The frame time histograms are divided in 1024 buckets, where each
```

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (dpallotti) [reply src/FakeClient/GraphicsCaptureEventProcessor.h:250]: Ok, I find it a bit strange though. At least add a comment that this number is arbitrary.

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (dpallotti) [inline src/FakeClient/GraphicsCaptureEventProcessor.h:231]: I suggest splitting this into two lines, one for the "\n" and one for the "%u,%.2f", to clarify where the generation of the header ends.

**Review by dpallotti — COMMENTED** [review]
Just some final minor stuff.

File: src/FakeClient/FakeCaptureEventProcessor.h
Comment (npcdoom) [reply src/FakeClient/FakeCaptureEventProcessor.h]: now for sure its fixed

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (npcdoom) [reply src/FakeClient/GraphicsCaptureEventProcessor.h]: done

File: src/FakeClient/FakeCaptureEventProcessor.h
Comment (npcdoom) [reply src/FakeClient/FakeCaptureEventProcessor.h]: done

File: src/FakeClient/FakeClientMain.cpp
Comment (npcdoom) [reply src/FakeClient/FakeClientMain.cpp]: done

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (npcdoom) [reply src/FakeClient/GraphicsCaptureEventProcessor.h]: done

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (npcdoom) [reply src/FakeClient/GraphicsCaptureEventProcessor.h]: done

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (npcdoom) [reply src/FakeClient/GraphicsCaptureEventProcessor.h]: done

File: src/FakeClient/FakeClientMain.cpp
Comment (npcdoom) [reply src/FakeClient/FakeClientMain.cpp]: done

File: src/FakeClient/GraphicsCaptureEventProcessor.h
Comment (npcdoom) [reply src/FakeClient/GraphicsCaptureEventProcessor.h]: done

File: src/FakeClient/FakeClientMain.cpp
Comment (dpallotti) [inline src/FakeClient/FakeClientMain.cpp:309]: Use the variable declared above.
```suggestion
    WaitForFileModification(pid_file_path);
    process_id = ReadPidFromFile(pid_file_path);
```

**Review by dpallotti — APPROVED** [review]
Note that there are a couple of very minor unresolved comments that you missed under "hidden conversations" (quite annoying from GitHub).

**Pull request merged**
